internal/biz/resources/resources.go

```diff
@@ -109,23 +111,125 @@ func (uc *Usecase) Create(ctx context.Context, m *model.Resource) (*model.Resour
 			if err != nil {
 				return nil, err
 			}
+
+			updatedResource.ConsistencyToken = ct
 		}
 	}
 
 	uc.log.WithContext(ctx).Infof("Created Resource: %v(%v)", m.ID, m.ResourceType)
 	return ret, nil
 }
 
+func (uc *Usecase) Check(ctx context.Context, permission, namespace string, sub *kessel.SubjectReference, id model.ReporterResourceId) (bool, error) {
+	res, err := uc.reporterResourceRepository.FindByReporterResourceId(ctx, id)
+	if err != nil {
+		if errors.Is(err, gorm.ErrRecordNotFound) {
+			// resource doesn't exist.
+			return false, nil
+		} else {
+			return false, err
+		}
+	}
+
+	allowed, _, err := uc.Authz.Check(ctx, namespace, permission, res, sub)
+	if err != nil {
+		return false, err
+	}
+
+	if allowed == kessel.CheckResponse_ALLOWED_TRUE {
+		return true, nil
+	} else {
+		return false, nil
+	}
+}
+
+func (uc *Usecase) CheckForUpdate(ctx context.Context, permission, namespace string, sub *kessel.SubjectReference, id model.ReporterResourceId) (bool, error) {
+	res, err := uc.reporterResourceRepository.FindByReporterResourceId(ctx, id)
+	if err != nil {
+		if errors.Is(err, gorm.ErrRecordNotFound) {
+			// resource doesn't exist yet.
+			res = &model.Resource{ResourceType: id.ResourceType, Reporter: model.ResourceReporter{LocalResourceId: id.LocalResourceId}}
+		} else {
+			return false, err
+		}
+	}
+
+	allowed, consistency, err := uc.Authz.CheckForUpdate(ctx, namespace, permission, res, sub)
+	if err != nil {
+		return false, err
+	}
+
+	if allowed == kessel.CheckForUpdateResponse_ALLOWED_TRUE {
+		if id.ResourceType == "workspace" && namespace == "rbac" { //TODO: delete this when workspaces are resources
+			return true, nil
+		}
+
+		if consistency != nil {
+			res.ConsistencyToken = consistency.Token
+			_, _, err := uc.reporterResourceRepository.Update(ctx, res, res.ID)
+			if err != nil {
+				return false, err
+			}
+		}
+
+		return true, nil
+	} else {
+		return false, nil
+	}
+}
+
+func (uc *Usecase) ListResourcesInWorkspace(ctx context.Context, permission, namespace string, sub *kessel.SubjectReference, id string) (chan *model.Resource, chan error, error) {
+	resource_chan := make(chan *model.Resource)
+	error_chan := make(chan error, 1)
+
+	resources, err := uc.reporterResourceRepository.FindByWorkspaceId(ctx, id)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	log.Infof("ListResourcesInWorkspace: resources %+v", resources)
+
+	go func() {
+		for _, resource := range resources {
+			log.Infof("ListResourcesInWorkspace: checkforview on %+v", resource)
+			if allowed, _, err := uc.Authz.Check(ctx, namespace, permission, resource, sub); err == nil && allowed == kessel.CheckResponse_ALLOWED_TRUE {
+				resource_chan <- resource
+			} else if err != nil {
+				error_chan <- err
+				break
+			} else if allowed != kessel.CheckResponse_ALLOWED_TRUE {
+				log.Infof("Response was not allowed: %v", allowed)
+			}
+
+		}
+
+		close(resource_chan)
+		close(error_chan)
```

Comment:
Any thoughts on using defer for these instead? This should give us some extra assurances for error cases